HARK/ConsumptionSaving/ConsRiskyAssetModel.py

```diff
@@ -0,0 +1,1944 @@
+    
+    
+    def simBirth(self,which_agents):
+        '''
+        Create new agents to replace ones who have recently died; takes draws of
+        initial aNrm and pLvl, as in ConsIndShockModel, then sets Share and Adjust
+        to zero as initial values.
+        Parameters
+        ----------
+        which_agents : np.array
+            Boolean array of size AgentCount indicating which agents should be "born".
+
+        Returns
+        -------
+        None
+        '''
+        
+        IndShockConsumerType.simBirth(self,which_agents)
+        self.state_now['nNrmTildeNow'][which_agents] = 0.
+        self.state_now['ShareNow'][which_agents] = 0.
+        # here a shock is being used as a 'post state'
+        self.shocks['AdjustNow'][which_agents] = False
+        
+            
+    def getShocks(self):
+        '''
+        Draw idiosyncratic income shocks, just as for IndShockConsumerType, then draw
+        a single common value for the risky asset return.  Also draws whether each
+        agent is able to update their risky asset share this period.
+        
+        Parameters
+        ----------
+        None
+
+        Returns
+        -------
+        None
+        '''
+        IndShockConsumerType.getShocks(self)
+        self.getRisky()
+        self.getAdjust()
+        
+    def simOnePeriod(self):
+        """
+        Simulates one period for this type.  Calls the methods getMortality(), getShocks() or
+        readShocks, getStates(), getControls(), and getPostStates().  These should be defined for
+        AgentType subclasses, except getMortality (define its components simDeath and simBirth
+        instead) and readShocks.
+        Parameters
+        ----------
+        None
+        Returns
+        -------
+        None
+        """
+        if not hasattr(self, "solution"):
+            raise Exception(
+                "Model instance does not have a solution stored. To simulate, it is necessary"
+                " to run the `solve()` method of the class first."
+            )
+                
+        # Mortality and birth happens only at the start of the period
+        self.getMortality()
+            
+        # Shocks are drawn in the first stage
+        if self.read_shocks:  # If shock histories have been pre-specified, use those
+            self.readShocks()
+        else:  # Otherwise, draw shocks as usual according to subclass-specific method
+            self.getShocks()
+        
+        # Stages in chronological order
+        stages = ['Reb','Sha','Cns']
+        
+        setStates = {'Reb': self.getStatesReb,
+                     'Sha': self.getStatesSha,
+                     'Cns': self.getStatesCns}
+        
+        setControls = {'Reb': self.getControlsReb,
+                       'Sha': self.getControlsSha,
+                       'Cns': self.getControlsCns}
+        
+        for s in stages:
+            setStates[s]()
+            setControls[s]()
```

Comment:
I'm curious about the motivation for this design.
Thus far we haven't been separating out the state and control transitions in simulation.
My understanding is that @mnwhite believes it is necessary to allow some model variables to be determined in the same method, for performance reasons.

Reply:
I did this when the frames conversation was just starting, so I had the `states -> controls -> postStates` framework in mind.

I am confident that when the frame tools are merged in I can make this (simulation) work with frames. Stages will still be important on the solution side, until generic frame-solution methods are added.